GPU/Debugger/Record.cpp

```diff
@@ -166,34 +168,44 @@ static const u8 *mymemmem(const u8 *haystack, size_t off, size_t hlen, const u8
 	}
 
 	const u8 *last_possible = haystack + hlen - nlen;
+	const u8 *first_possible = haystack + off;
 	int first = *needle;
-	const u8 *p = haystack + off;
 
-	const uintptr_t align_mask = align - 1;
-	auto poffset = [&]() {
-		return ((uintptr_t)(p - haystack) & align_mask);
-	};
-	auto alignp = [&]() {
-		uintptr_t offset = poffset();
-		if (offset != 0)
-			p += align - offset;
-	};
+	std::atomic<const u8 *> result;
+	result.store(nullptr);
 
-	alignp();
-	while (p <= last_possible) {
-		p = (const u8 *)memchr(p, first, last_possible - p + 1);
-		if (!p) {
-			return nullptr;
-		}
-		if (poffset() == 0 && !memcmp(p, needle, nlen)) {
-			return p;
-		}
+	int range = (int)(last_possible - first_possible);
+	GlobalThreadPool::Loop([&](int l, int h) {
+		const u8 *p = haystack + off + l;
+		const u8 *pend = haystack + off + h;
+
+		const uintptr_t align_mask = align - 1;
+		auto poffset = [&]() {
+			return ((uintptr_t)(p - haystack) & align_mask);
+		};
+		auto alignp = [&]() {
+			uintptr_t offset = poffset();
+			if (offset != 0)
+				p += align - offset;
+		};
 
-		p++;
 		alignp();
-	}
+		while (p <= pend && !result.load()) {
+			p = (const u8 *)memchr(p, first, pend - p + 1);
+			if (!p) {
+				return;
+			}
+			if (poffset() == 0 && !memcmp(p, needle, nlen)) {
+				result.store(p);
```

Comment:
So the first thread that finds a result "wins", and it's no longer deterministic that it's the first one?

All the other changes in thie PR are easy to reason about and I'm happy to merge, but this one feels a little scary, I have to say.

Reply:
Hm, it really only needs to be any (it's essentially an unlimited backreference), but probably better to make it deterministic indeed.  Really, the most efficient thing might be for it to always be the *last* one, but I don't want to have to search through all the memory to find it.

-[Unknown]

Reply:
oh right, this is for compression. Still, determinism feels better :)

Reply:
Yep, agreed.  Sorry got sidetracked and forgot about this temporarily.  Updated.

-[Unknown]